### PR TITLE
Create order

### PR DESCRIPTION
**Pull request by ConfusedSatlan**

### CODE REVIEW

**Review by Zel1oy — APPROVED** [review]
Good solution!

File: src/main/java/application/onlinebookstore/controller/OrdersController.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+@Tag(name = "Orders management", description = "Endpoints for managing orders")
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/api/orders")
+public class OrdersController {
+    private final OrdersService ordersService;
+    private final OrderItemService orderItemService;
+
+    @PreAuthorize("hasRole('ROLE_USER')")
```
Comment (vovasalyha) [inline src/main/java/application/onlinebookstore/controller/OrdersController.java]: I think admin also should have access to GET methods

File: src/main/java/application/onlinebookstore/model/Orders.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    private String shippingAddress;
+    @Enumerated(EnumType.STRING)
+    @JoinColumn(name = "status",nullable = false)
+    private Status status;
+    @JoinColumn(name = "total",nullable = false)
+    private BigDecimal total;
+    @Column(name = "is_deleted", nullable = false)
+    private boolean isDeleted = false;
+
+    public enum Status {
```
Comment (vovasalyha) [inline src/main/java/application/onlinebookstore/model/Orders.java:63]: I think COMPLETED and NOT_COMPLETED is not all of the statuses. Google possible order statuses for example store and add more (for example CANCELLED)

File: src/main/java/application/onlinebookstore/model/Orders.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+@Data
+@Entity
+@SQLDelete(sql = "UPDATE orders SET is_deleted = true WHERE id=?")
+@Where(clause = "is_deleted=false")
+@Table(name = "orders")
+public class Orders {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne
```
Comment (vovasalyha) [inline src/main/java/application/onlinebookstore/model/Orders.java]: why not lazy?

File: src/main/java/application/onlinebookstore/model/Orders.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    @JoinColumn(name = "user_id")
+    @ToString.Exclude
+    @EqualsAndHashCode.Exclude
+    private User user;
+    @OneToMany(fetch = FetchType.EAGER)
```
Comment (vovasalyha) [inline src/main/java/application/onlinebookstore/model/Orders.java]: why not lazy?

File: src/main/java/application/onlinebookstore/repository/orderitem/OrderItemRepository.java [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
+import org.springframework.data.jpa.repository.Query;
+
+public interface OrderItemRepository extends JpaRepository<OrderItem, Long>,
+        JpaSpecificationExecutor<OrderItem> {
+    List<OrderItem> findAllByOrderId(Long id);
+
+    @Query("SELECT oi FROM OrderItem oi WHERE oi.id = :id AND oi.order.id = :orderId")
+    OrderItem findByIdAndOrderId(Long id, Long orderId);
```
Comment (vovasalyha) [inline src/main/java/application/onlinebookstore/repository/orderitem/OrderItemRepository.java:14]: it should work without custom @Query if I'm not mistaken

File: src/main/java/application/onlinebookstore/model/Orders.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+import org.hibernate.annotations.CascadeType;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
+
+@Data
+@Entity
+@SQLDelete(sql = "UPDATE orders SET is_deleted = true WHERE id=?")
+@Where(clause = "is_deleted=false")
+@Table(name = "orders")
+public class Orders {
```
Comment (vovasalyha) [inline src/main/java/application/onlinebookstore/model/Orders.java:32]:
```suggestion
public class Order {

```

File: src/main/java/application/onlinebookstore/service/OrderItemService.java [hunk 7]
```diff
@@ -0,0 +1,13 @@
+package application.onlinebookstore.service;
+
+import application.onlinebookstore.dto.orderitem.CreateOrderItemDto;
+import application.onlinebookstore.dto.orderitem.OrderItemDto;
+import java.util.List;
+
+public interface OrderItemService {
+    OrderItemDto createOrderItem(CreateOrderItemDto orderItemDto);
```
Comment (vovasalyha) [inline src/main/java/application/onlinebookstore/service/OrderItemService.java:8]:
```suggestion
    OrderItemDto create(CreateOrderItemDto orderItemDto);

```
IMO OrderItem postfix can be omitted in OrderItemService, but it's for you to decide

File: src/main/java/application/onlinebookstore/service/OrdersService.java [hunk 8]
```diff
@@ -0,0 +1,15 @@
+package application.onlinebookstore.service;
+
+import application.onlinebookstore.dto.orders.CreateOrderDto;
+import application.onlinebookstore.dto.orders.OrderDto;
+import application.onlinebookstore.model.Orders;
+import java.util.List;
+
+public interface OrdersService {
```
Comment (vovasalyha) [inline src/main/java/application/onlinebookstore/service/OrdersService.java:8]:
```suggestion
public interface OrderService {

```

File: src/main/java/application/onlinebookstore/service/impl/OrderItemServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    public OrderItemServiceImpl(OrderItemRepository orderItemRepository,
+                                BookService bookService,
+                                OrderItemMapper orderItemMapper) {
+        this.orderItemRepository = orderItemRepository;
+        this.bookService = bookService;
+        this.orderItemMapper = orderItemMapper;
+    }
+
+    @Autowired
+    public void setOrdersService(@Lazy OrdersService ordersService) {
```
Comment (vovasalyha) [inline src/main/java/application/onlinebookstore/service/impl/OrderItemServiceImpl.java:35]: Why it should be lazy?

File: src/main/java/application/onlinebookstore/service/impl/OrdersServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    private final OrderMapper orderMapper;
+
+    @Override
+    public List<OrderDto> getOrders(Long userId) {
+        return ordersRepository.findAllByUserId(userId).stream()
+                .map(orderMapper::toDto)
+                .toList();
+    }
+
+    @Override
```
Comment (vovasalyha) [inline src/main/java/application/onlinebookstore/service/impl/OrdersServiceImpl.java:46]:
```suggestion
    @Override
    @Transactional
```
Think carefully which methods need to be transactional

File: src/main/java/application/onlinebookstore/service/impl/OrdersServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+    public OrderDto create(CreateOrderDto orderDto, Long userId) {
+        User userById = userService.getUserById(userId);
+        Orders order = new Orders();
+        order.setUser(userById);
+        order.setOrderItems(new HashSet<>());
+        order.setOrderDate(LocalDateTime.now());
+        order.setShippingAddress(orderDto.shippingAddress());
+        order.setStatus(Orders.Status.NOT_COMPLETED);
+        order.setTotal(new BigDecimal(0));
```
Comment (vovasalyha) [inline src/main/java/application/onlinebookstore/service/impl/OrdersServiceImpl.java]:
```suggestion
        order.setTotal(BigDecimal.ZERO);

```

File: src/main/java/application/onlinebookstore/service/impl/OrdersServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+        ShoppingCartDto shoppingCart = shoppingCartService.getShoppingCart(userId);
+        Set<CartItemDto> cartItems = shoppingCart.getCartItems();
+        BigDecimal total = new BigDecimal(0);
+        for (CartItemDto cartItem : cartItems) {
+            CreateOrderItemDto orderItemDtoParams = new CreateOrderItemDto(
+                    order.getId(),
+                    cartItem.getBookId(),
+                    cartItem.getQuantity()
+            );
+            OrderItemDto orderItemDto = orderItemService.createOrderItem(orderItemDtoParams);
```
Comment (vovasalyha) [inline src/main/java/application/onlinebookstore/service/impl/OrdersServiceImpl.java:68]: it's not necessary to call save each time, when you're calling `order.getOrderItems().add()` and then save an order entity, all order items will be saved too, even in more productive way by using batch insert

File: src/main/resources/db/changelog/changes/17-create-order-item-table.yaml [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+                  type: bigint
+                  autoIncrement: true
+                  constraints:
+                    primaryKey: true
+              - column:
+                  name: orders_id
+                  type: bigint
+                  constraints:
+                    nullable: false
+                    foreignKeyName: orders_id_fk
```
Comment (vovasalyha) [inline src/main/resources/db/changelog/changes/17-create-order-item-table.yaml]: and to which table and column it references? specify param `references` for all foreign keys

**Review by vovasalyha — CHANGES_REQUESTED** [review]
Good, left some comments

File: src/main/java/application/onlinebookstore/dto/orders/CreateOrderDto.java [hunk 14]
```diff
@@ -0,0 +1,4 @@
+package application.onlinebookstore.dto.orders;
+
+public record CreateOrderDto(String shippingAddress) {
```
Comment (andrii-hoienko) [inline src/main/java/application/onlinebookstore/dto/orders/CreateOrderDto.java]: Maybe it makes sense to add some validation here

File: src/main/java/application/onlinebookstore/model/Orders.java (hunk 3 above)
Comment (andrii-hoienko) [reply src/main/java/application/onlinebookstore/model/Orders.java]: Not resolved

File: src/main/java/application/onlinebookstore/model/Orders.java (hunk 2 above)
Comment (andrii-hoienko) [reply src/main/java/application/onlinebookstore/model/Orders.java:63]: Not resolved

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Please resolve all comments

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Please read all comments provided my mentors once more

**Review by vovasalyha — APPROVED** [review]

**Pull request merged**